R/iterate-helpers.R

```diff
@@ -0,0 +1,82 @@
+#' Iteration helpers
+#'
+#' @description
+#' These functions are suitable for use with the `next_req` argument to
+#' [req_perform_iteratively()]. Each implements iteration for a common
+#' pagination pattern:
+#'
+#' * `iterate_with_offset()` increments a query parameter, e.g. `?page=1`,
+#'   `?page=2`, or `?offset=1`, `offset=21`.
+#' * `iterate_with_cursor()` updates a query parameter with the value of a
+#'    cursor found somewhere in the response.
+#'  * `iterate_with_link_url()` follows the url found the `Link` header.
+#'    See `resp_link_url()` for more details.
+#'
+#' @param param_name Name of query parameter.
+#' @param start Starting value.
+#' @param offset Offset for each page.
+#' @param resp_complete A callback function that takes a response (`resp`)
+#'   and returns `TRUE` if there are no further pages.
+#' @export
+#' @examples
+#' req <- request(example_url()) |>
+#'   req_url_path("/iris") |>
+#'   req_throttle(10) |>
+#'   req_url_query(limit = 5)
+#'
+#' is_complete <- function(resp) {
+#'   length(resp_body_json(resp)$data) == 0
+#' }
+#' resps <- req_perform_iteratively(
+#'   req,
+#'   iterate_with_offset("page_index", resp_complete = is_complete)
+#' )
+iterate_with_offset <- function(param_name,
+                                start = 1,
+                                offset = 1,
+                                resp_complete = NULL) {
```

Comment:
`resp_complete()` is a different way to handle the maximum number of pages to request issue.